core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -0,0 +1,307 @@
+import io.grpc.Compressor;
+import io.grpc.DecompressorRegistry;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+
+import java.io.InputStream;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.Executor;
+
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * A stream that queues requests before the transport is available, and delegates to a real stream
+ * implementation when the transport is available.
+ *
+ * <p>{@code ClientStream} itself doesn't require thread-safety. However, the state of {@code
+ * DelayedStream} may be internally altered by different threads, thus internal synchronization is
+ * necessary.
+ */
+final class DelayedStream implements ClientStream {
+  private final Executor callExecutor;
+  private final MethodDescriptor<?, ?> method;
+
+  private final Metadata headers;
+  private final ClientStreamListener listener;
+  private final CallOptions callOptions;
+
+  private final Object lock = new Object();
+
+  // Volatile to be readable without synchronization in the fast path.
+  // Writes are also done within synchronized(this).
+  private volatile ClientStream realStream;
+
+  @GuardedBy("lock")
+  private Compressor compressor;
+  // Can be either a Decompressor or a String
+  @GuardedBy("lock")
+  private Object decompressor;
+  @GuardedBy("lock")
+  private DecompressorRegistry decompressionRegistry;
+  @GuardedBy("lock")
+  private final List<PendingMessage> pendingMessages = new LinkedList<PendingMessage>();
+  private boolean messageCompressionEnabled;
+  @GuardedBy("lock")
+  private boolean pendingHalfClose;
+  @GuardedBy("lock")
+  private int pendingFlowControlRequests;
+  @GuardedBy("lock")
+  private boolean pendingFlush;
+
+  static final class PendingMessage {
+    final InputStream message;
+    final boolean shouldBeCompressed;
+
+    public PendingMessage(InputStream message, boolean shouldBeCompressed) {
+      this.message = message;
+      this.shouldBeCompressed = shouldBeCompressed;
+    }
+  }
+
+  DelayedStream(
+      CallOptions callOptions,
+      Metadata headers,
+      ClientStreamListener listener,
+      Executor callExecutor,
+      MethodDescriptor<?, ?> method) {
+    this.headers = headers;
+    this.listener = listener;
+    this.callExecutor = callExecutor;
+    this.method = method;
+    this.callOptions = callOptions;
+  }
+
+  /**
+   * Creates a stream on a presumably usable transport.
+   */
+  void createStream(ClientTransport transport) {
+    synchronized (lock) {
+      if (realStream == NOOP_CLIENT_STREAM) {
+        // Already cancelled
+        return;
+      }
+      checkState(realStream == null, "Stream already created: %s", realStream);
+      if (!ClientCallImpl.updateTimeoutHeader(callOptions.getDeadlineNanoTime(), headers)) {
+        maybeClosePrematurely(Status.DEADLINE_EXCEEDED);
+        return;
+      }
+      realStream = transport.newStream(method, headers, listener);
```

Comment:
There isn't need for a lock when doing `updateTimeoutHeader()` because only that one thread will be using Metadata (and if that wasn't the case, then the `DelayedStream` lock would be no better). There is really no way to use gRPC while retaining a reference to the Metadata passed to `start()`, because the object is not thread-safe.

I don't mind too much having `maybeClosePrematurely()` visible, simply because it seems a bit odd already. `maybeClosePrematurely()` is the only use of `listener` in `DelayedStream`. Yet the method doesn't work well in `ClientCallImpl` either because of `DelayedStream.cancel()`. I'm thinking the oddity will go away when we separate `start()` from `newStream()`. Something for another day.